### DISCUSSION TITLE
Fetch data for component during build time only

### DISCUSSION THREAD

**Discussion by stevenmunro** [vercel/next.js#16668] [Help]
I suppose what I'm looking for is something like `getStaticProps`, but for components. Unfortunately `getStaticProps` only works for pages.

Say I have inside my component a reference to an image. `<img src={require('../public/photo.jpg')} />`. This image is **required** at build time, otherwise you would get a `module not found` error.

I wish to replace the src of this image with a base64 encoded data-uri. This data-uri comes from a third party service (visit url, returns the string); But I don't want the data-uri to be fetched each time the component is mounted. This is a 20x20px pixel image used for blur-up effect, so it really needs to be part of the page during the time of request.

Does anyone know the best way to achieve this?

(edited for clarity)

**Comment by franky47** [comment]
You could use `getStaticProps` in the pages where your component might live, and use some form of Context to pass static data to your components (or prop drilling, depending on what's least inconvenient). This would be using only Next.js/React features.

I had a similar use-case for my website, where I wanted to embed tweets, Unsplash images (similar use-case as yours I believe), Spotify albums, GitHub repository metadata etc. I did not want to tie every single page's data fetching to components data requirements, so I settled for fetching the data in a pre-build phase, export it as .js/.ts files[1], then `require` them to feed the view components.

I might write an article on this, in the mean time here's a link to the implementation:
https://github.com/franky47/francoisbest.com/tree/next/src/data

[1] .js/.ts instead of .json as `require('xyz.json')` are not code-splitted, but JS/TS files are. This avoids shipping all your static component data to any page that uses at least one.
